### ISSUE TITLE
Simplifying the model creation setup.

### ISSUE THREAD

**Issue by davidorme** [ImperialCollegeLondon/virtual_ecosystem#369]
This is a proposal for a refactor of the BaseModel to make a easier to update interface 
between the core and specific models.

We are in the process of adding new core content into models, centralising 
a bunch of stuff. I've done that for core_constants in a WIP pull request (#368), 
but it was a surprising pain in the arse, because:

* a `core_constants` instance first needs to be created in `vr_run` before models 
  are initialised.
* the argument then needs to be added to the `BaseModel.__init__` signature...
* ... and every model's `__init__` signature...
* ... and every model's `from_config` signature.

There is a bunch of core stuff on its way:

* core constants
* vertical layer structure
* axis definitions
* temporal definitions

It makes a lot of sense to add this all to model objects as attributes using `BaseModel.__init__()`, so
that the required `super().__init__()` call within a model populates all that stuff. The simplest thing
to avoid complex signatures, is simply to pass it a `Config` object and use that to
populate those base attributes

But - now we have a model `__init__` signature that contains `Config`, so what is the point
in the `from_config method`. I think we had a split between programmatic creation of a
model instance - where a detailed signature might help people understand model
differences - and then the runtime creation in simulation, where we just supply the
config.

I'm  not sure that keeping that model specific signature overrides the complexity needed
to add new attributes. We could just get rid of from_config approach and let all models 
be defined from simple core inputs. Models could then basically be created using:

```python
model <- ModelClass(data=data, config=config)
```
The `__init__` method - or possibly an encapsulated `configure` abstract method then
checks if the model can be configured from the inputs.

As a simplified sketch:

```python
from abc import ABC

def build_widely_used_thing(thing: str) -> str:
    """Build a widely used thing from core config contents."""
    return thing

class DemoConfig(dict):
    """Demo config."""
    pass

class DemoBase(ABC):
    """A base class that sets a bunch of widely used stuff directly from the config."""
    def __init__(self, config: DemoConfig) -> None:
        self.config = config
        self.widely_used_thing = build_widely_used_thing(config["core_thing"])

class DemoModel(DemoBase):
    """A daughter of the base class that adds a new model_thing attribute."""
    def __init__(self, config: DemoConfig):
        super().__init__(config)
        self.model_thing = self.config["model_thing"]
        print(self.__class__)

cfg = DemoConfig(
    {
        "layers": [1, 2, 3, 4],
        "core_thing": "core_thing",
        "model_thing": "model_thing",
    }
)

model = DemoModel(config=cfg)
print(model.widely_used_thing)
print(model.model_thing)
print(model.config["layers"])
```

Using this approach, we can expose new core components by simply editing `BaseModel.__init__()` to add the attribute in one place.

Another issue is that the approach above creates a new copy of `widely_used_thing` in each model and we might want to add the creation of those things to - for example - `Config`, which gets created once and then `BaseModel` can unpack whichever bits are needed (or leave them accessible via `self.config` if they are less commonly used?).

```python
class DemoConfig2(dict):
    """Demo config with core components."""

    def __init__(self, input: dict) -> None:
        """Build shared things once in config."""
        super().__init__(input)
        self.widey_used_thing: str = build_widely_used_thing(input["core_thing"])

class DemoBase2(ABC):
    """A base class that sets a bunch of widely used stuff directly from the config."""
    def __init__(self, config: DemoConfig2) -> None:
        self.config = config
        self.widely_used_thing = config.widey_used_thing

class DemoModel2(DemoBase2):
    """A daughter of the base class that adds a new model_thing attribute."""
    def __init__(self, config: DemoConfig2):
        super().__init__(config)
        self.model_thing = self.config["model_thing"]

cfg2 = DemoConfig2(
    {
        "layers": [1, 2, 3, 4],
        "core_thing": "core_thing",
        "model_thing": "model_thing",
    }
)

model2 = DemoModel2(config=cfg2)

print(model2.widely_used_thing)
print(model2.model_thing)
print(model2.config["layers"])
```

**Comment by dalonsoa** [comment]
I think it is a pity to loose the specificity of the model signatures. It is useful documentation in itself, much better than an abstract dictionary that you need to explain. Maybe a midpoint will be to have your second approach but mixed with the `from_config` one. Something like:

```python
class DemoConfig2(dict):
    """Demo config with core components."""

    def __init__(self, input: dict) -> None:
        """Build shared things once in config."""
        super().__init__(input)
        self.widey_used_thing: str = build_widely_used_thing(input["core_thing"])

class DemoModel3(AppropriateBaseModel):
    """A daughter of the base class built from the configuration."""
    @classmethod  
    def from_config(self, config: DemoConfig2) -> None:
        model_thing = self.config["model_thing"]
        widely_used_thing = config.widey_used_thing
        return cls(model_thing, widely_used_thing)

    def __init__(self, model_thing: int, widely_used_thing: str):
        self.model_thing = model_thing
        self.widely_used_thing = widely_used_thing

cfg2 = DemoConfig2(
    {
        "layers": [1, 2, 3, 4],
        "core_thing": "core_thing",
        "model_thing": "model_thing",
    }
)

model3 = DemoModel3.from_config(config=cfg2)

print(model3.widely_used_thing)
print(model3.model_thing)
```
In this scenario, you keep the specificity of the signature and do not add ALL, mostly unnecessary, possible configuration information to the classes, just what they need to work.

**Comment by davidorme** [comment]
I agree that the model specific signatures is a loss. Your alternative keeps the model attributes slimmed down, which I sort of think is a good thing, although having common resources in all models seems like a more integrated approach. 

That doesn't really resolve the underlying issue though, which is that each core element is being introduced in a piecemeal fashion in three locations, rather than being defined in one location and inherited via the base class. If we want to keep using the ABC for BaseModel (which is pretty fundamental and I don't think we should dump!) then we do have this constraint about what we can populate _within_ the `BaseModel.__init__()` and that seems very inelegant.


-  unless we can add core attributes via `BaseModel.__init__()`then we're left with having to add new attributes and signature elements for many core attributes. I don't think

**Comment by davidorme** [comment]
I've had a look at the current model signatures (actually, the versions of them including the core_constants from #368). There is a large proportion of core elements, all of which are being recreated or extracted independently in every `from_config` and passed to `__init__` and which could be inherited directly via `super().__init__()` if we switch to the structure above.

```python
class PlantsModel(BaseModel, ...):
    def __init__(
        self,
        data: Data,                                      # CORE
        core_constants: CoreConsts,                      # CORE
        update_interval: Quantity,                       # CORE
        flora: Flora,                                    # MODEL
        layer_structure: LayerStructure,                 # CORE
        model_constants: PlantsConsts = PlantsConsts(),  # MODEL
        **kwargs: Any,
    ):

class SoilModel(BaseModel, ...):
    def __init__(
        self,
        data: Data,                                      # CORE
        core_constants: CoreConsts,                      # CORE
        update_interval: Quantity,                       # CORE
        soil_layers: list[float],                        # CORE
        canopy_layers: int,                              # CORE
        model_constants: SoilConsts,                     # MODEL
        **kwargs: Any,
    ):


class LitterModel(BaseModel, ...):
    def __init__(
        self,
        data: Data,                                      # CORE
        core_constants: CoreConsts,                      # CORE
        update_interval: Quantity,                       # CORE
        soil_layers: list[float],                        # CORE
        canopy_layers: int,                              # CORE
        model_constants: LitterConsts,                   # MODEL
        **kwargs: Any,
    ):
        

class LitterModel(BaseModel, ...):
    def __init__(
        self,
        data: Data,                                      # CORE
        core_constants: CoreConsts,                      # CORE
        update_interval: Quantity,                       # CORE
        soil_layers: list[float],                        # CORE
        canopy_layers: int,                              # CORE
        initial_soil_moisture: float,                    # MODEL
        initial_groundwater_saturation: float,           # MODEL
        constants: HydroConsts,                          # MODEL
        **kwargs: Any,
    ):

class AnimalModel(BaseModel, ...):
    def __init__(
        self,
        data: Data,                                      # CORE
        core_constants: CoreConsts,                      # CORE
        update_interval: Quantity,                       # CORE
        functional_groups: list[FunctionalGroup],        # MODEL
        model_constants: AnimalConsts = AnimalConsts(),  # MODEL
        **kwargs: Any,
    ):

class AbioticSimpleModel(BaseModel, ...):
    def __init__(
        self,
        data: Data,                                      # CORE
        core_constants: CoreConsts,                      # CORE
        update_interval: Quantity,                       # CORE
        soil_layers: list[float],                        # CORE
        canopy_layers: int,                              # CORE
        constants: AbioticSimpleConsts,                  # MODEL
        **kwargs: Any,
    ):


class AbioticModel(BaseModel, ...):
    def __init__(
        self,
        data: Data,                                      # CORE
        update_interval: Quantity,                       # CORE
        soil_layers: list[float],                        # CORE
        canopy_layers: int,                              # CORE
        constants: AbioticConsts,                        # MODEL
        core_constants: CoreConsts,                      # CORE
        **kwargs: Any,
    ):
```

**Comment by dalonsoa** [comment]
Go for it, then. I'm not super enthusiastic about having a dictionary (even with some structure) as input as the signature becomes blurred, but I agree it will make the definition of the models much simpler. And I cannot think on a better alternative, to be honest.

**Comment by davidorme** [comment]
I can see why you don't like that collapsed signature - I can't see an easy way to expose the model specific elements. We could create a `configure` method but that either looks like `configure(config = config)` or `configure(thing = config['bits']['for']['thing'], thing2 = config['bits']['for']['thing'], ...)`  and then that assumes lot about the keys. We could simply have model `__init__` use:

```python
try:
    self.configure(thing = config['bits']['for']['thing'], thing2 = config['bits']['for']['thing'], ...)
except KeyError as excep:
    logger.Error('Missing configuration element') # Although the schema validation should defend against this!
    raise ConfigurationError()
```

That does allow us to have something within each model that documents the model specific elements?

**Comment by dalonsoa** [comment]
Is there a reason why we cannot do something like:

```python
class DemoConfig2(dict):
    """Demo config with core components."""

    def __init__(self, input: dict) -> None:
        """Build shared things once in config."""
        super().__init__(input)
        self.widey_used_thing: str = build_widely_used_thing(input["core_thing"])

class DemoBase2(ABC):
    """A base class that sets a bunch of widely used stuff directly from the config."""
    def __init__(self, config: DemoConfig2) -> None:
        self.config = config
        self.widely_used_thing = config.widey_used_thing

class DemoModel2(DemoBase2):
    """A daughter of the base class that adds a new model_thing attribute."""

    @classmethod  
    def from_config(cls, config: DemoConfig2) -> None:
        model_thing = config["model_thing"]
        return cls(model_thing, config)

    def __init__(self, model_thing: str, config: DemoConfig2):
        super().__init__(config)
        self.model_thing = model_thing

cfg2 = DemoConfig2(
    {
        "layers": [1, 2, 3, 4],
        "core_thing": "core_thing",
        "model_thing": "model_thing",
    }
)

model2 = DemoModel2(config=cfg2)

print(model2.widely_used_thing)
print(model2.model_thing)
print(model2.config["layers"])
```

Such that the model-specific fields are still unpacked, validated and whatever else is relevant in a `from_config` method with an explicit signature for the model-specific stuff, but we keep cascading the `config` down to the base class?

It might be overcomplicating things, to be honest. Your solution seems somwhat cleaner despite loosing the signature.

**Comment by davidorme** [comment]
I think we could definitely do that. 

* The downside is that all the information used in `DemoModel.from_config` is right there in `Demo_Model.__init__`, so it does feel like an unnecessary side step.
* The upsides are that i) it retains the model specific inputs and classes in the `__init__` signature, ii) it provides direct `BaseModel` inheritance and iii) it is fairly simple to go from this intermediate solution to the fully collapsed signatures - we'd just merge `from_config` into `__init__` (or some kind of encapsulated `configure` step).

**Comment by davidorme** [comment]
@dalonsoa I had a thought - for basically all of these core elements, what we really want to do is create them once and then pass by reference down through the models. I'd been thinking of adding these as attributes of `Config` but that basically makes `Config` into a waste-paper basket of an object. It would be better to instead create a `CoreComponents` dataclass, whose sole job is to contain single instances of universal core structures that models need.  

That allows us to retain the `from_config` factory method and have clear model specific signatures. Something like:

``` python

class DemoConfig2(dict):
    """Demo config."""

    pass


@dataclass
class CoreComponents:
    """A core component collection."""

    widely_used_thing: str = field(init=False)
    config: InitVar[DemoConfig2]

    def __post_init__(self, config: DemoConfig2) -> None:
        """Populate the core components from the config."""
        self.widely_used_thing = build_widely_used_thing(config["core_thing"])


class DemoBase2(ABC):
    """A base class that sets a bunch of widely used stuff directly from the config."""

    def __init__(self, core_components: CoreComponents) -> None:
        self.widely_used_thing = core_components.widely_used_thing


class DemoModel2(DemoBase2):
    """A daughter of the base class that adds a new model_thing attribute."""

    def __init__(self, core_components: CoreComponents, model_thing: str):
        super().__init__(core_components=core_components)
        self.model_thing = model_thing

    @classmethod
    def from_config(
        cls, core_components: CoreComponents, config: DemoConfig2
    ) -> DemoModel2:
        """Class factory for the model."""

        return cls(core_components=core_components, model_thing=config["model_thing"])


cfg = DemoConfig2(
    {
        "layers": [1, 2, 3, 4],
        "core_thing": "core_thing",
        "model_thing": "model_thing",
    }
)

core_components = CoreComponents(cfg)

model = DemoModel2.from_config(core_components=core_components, config=cfg)


print(model.widely_used_thing)
print(model.model_thing)
```

We could still in the future collapse this down into `DemoModel(core_components, config)` and merge the `from_config` into `__init__` but this feels like a better place to move to. Thoughts?

**Comment by dalonsoa** [comment]
I thought about this, but did not want to suggest yet another class. If you are happy with this, I think this would be the best option.

**Comment by davidorme** [comment]
Is there a problem with another class? Or you just thought we'd react badly 😄

**Comment by dalonsoa** [comment]
More the second :P

**Comment by davidorme** [comment]
Closed by #374 and #375
